lib/api/ldp/api.js

```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+const LdpRequest = require('./ldp-request')
+const AclChecker = require('../acl-checker')
+const LdpFileStore = require('../storage/ldp-file-store')
+
+class LdpApi {
+  /**
+   * @param options.host {SolidHost} server host / config object
+   *
+   * @param options.store {LdpFileStore|LdpMemoryStore|LdpQuadStore} storage backend
+   *   (either file system based, or in-memory, or SPARQL based quad store, etc)
+   *
+   * @param options.fetch {fetch} whatwg fetch, possibly with delegation in the
+   *   future. Needed for COPY operations, etc.
+   *
+   * @param options.acl {AclChecker} ACL checker component
+   */
+  constructor (options) {
+    this.host = options.host
+    this.fetch = options.fetch
+    this.store = options.store || new LdpFileStore(options)
+    this.acl = options.acl ||
+      new AclChecker({host: this.host, store: this.store, fetch: this.fetch})
+  }
+
+  /**
+   * Provides an Express handler for LDP API requests.
+   * Authentication has already happened earlier in the middleware stack, stored
+   * in session.
+   * Usage:
+   *   ```
+   *   app.use('/', ldp.middleware)
+   *   ```
+   *
+   * @see https://github.com/solid/solid-architecture/blob/master/server/request-flow.md
+   *
+   * @throws {Error} 401 If request not authenticated but resource is non-public
+   *
+   * @throws {Error} 404 If resource is not found, OR if found and request
+   *   is not authorized to access it (default behavior, can be overridden by
+   *   owner of resource)
+   *
+   * @throws {Error} 403 If request is not authorized to access resource and
+   *   user has enabled "Request permission" action
+   *
+   * @throws {Error} 400 If invalid parameters (or error parsing request body,
+   *   for cases like PATCH requests)
+   *
+   * @throws {Error} 409 If PATCH operation results in conflict
+   *
+   * @throws {Error} 406 If no appropriate representation found (for content type)
+   *
+   * @throws {Error} 405 If HTTP method not allowed / not implemented
+   */
+  async middleware (req, res, next) {
+    try {
+      // parse target, operation, (optionally) body, preferences, store authn
+      const request = LdpRequest.from({req, host: this.host})
```

Comment:
The extra host parameter feels a bit clumsy; it's also incomplete because the protocol is not considered. Maybe we should more generally have an `overrides` parameter or similar.

Reply:
how about if we call it `serverConfig` instead of `host`?

Reply:
The name makes sense.

But actually, I think you just need to parse `target` first as follows:
- have a `resourceMapper` private field, which is an instance of `https://github.com/solid/node-solid-server/blob/v4.0.14/lib/resource-mapper.js` (this already contains all the overrides)
- set `const target = this.resourceMapper.mapUrlToFile(…)`
- pass `target` to the LDP request constructor.

Alternatively, the `resourceMapper` can be passed to the LDP request constructor.

Reply:
Re 'private field' - heh heh, so, I don't really believe in private fields. I think Steve Yegge captures it perfectly with his essay: http://steve-yegge.blogspot.com/2010/07/wikileaks-to-leak-5000-open-source-java.html)

But that aside - so here's what's interesting about the `resourceMapper` (which I agree is useful).
It belongs to the *backend*, not to the API request, specifically only to the `LdpFileStore` (and not to an in-memory store or a Quad Store backend).

So the idea here is that the Request objects would only need to call `this.store.exists(resource)` and have the backend (if appropriate) worry about file mapping.

Reply:
> Actually, I think that LdpApi should take an ldpProvider as argument.

That's exactly what the `store` argument is, it's an ldp provider.

Reply:
Would you like it to be renamed to that?

Reply:
Ah right. I don't mind the name too much, I just think that `host` and  `fetch` should be dropped as parameters of `LdpApi`; they are the responsibility of the store (since all `LdpApi` does is passing them). Law of Demeter.

Reply:
And there should just be no defaults for `store` and `acl`; the API shouldn't need to know how to construct them. All the API should do is send the five steps the right way.

So if I'm very critical/demanding, I think that AclApi should almost literally reflect https://github.com/solid/solid-architecture/blob/master/server/request-flow.md: get all the right variables, and pass them to other components. That's more or less all it should be doing. Then it's up to those components to get instantiated with what they need (server options etc.) to fulfill their task.

Reply:
Ok. Will keep the `store` name.

I wasn't sure about passing in the `host` and `fetch` as separate parameters, either.
But they're not just needed in the store. The `host` and `fetch` are both needed by the ACL Checker. (And I was initially thinking that the `host` may be needed in the url parsing of `target`,  but maybe not).

I suppose I can not pass them in, and just require passing in an already-initialized `acl` checker object (that will already have them)

Reply:
So something like:

```
const target = this.targetParser.parse(request);
const operation = this.operationParser.parse(request);
const bodyParser = this.bodyParsers.find(request.headers['content-type']);
const body = bodyParser && bodyParser.parse(body);
```

Reply:
> And there should just be no defaults for store and acl

Ok, copy that. Will remove :)

And hey, critical & demanding is perfect, that's what I want :) I'm happy to have LdpApi reflect the request-flow document (with some slight exceptions).

Reply:
Ok, so let's take those one at a time.

> `const operation = this.operationParser.parse(request)`

Is the `operationParser()` meant to take the place of the `switch()` statement? Cause that's essentially what it is, an operations parsing & factory method. 

> `const target = this.targetParser.parse(request)`

This makes sense, though I'm somewhat unclear on if we need any complicated parsing for the target url. (Which is different from url-to-fs mapping, a domain of the backend).

> `const bodyParser = this.bodyParsers.find(request.headers['content-type'])`

I feel that this belongs in each operation subclass. Like, HEAD ops don't need parsing. Neither do PUTs and POSTS etc. But PATCH does.

Reply:
> Is the operationParser() meant to take the place of the switch() statement?

No:

>> The operation consists of the HTTP method, and a set of flags. Valid operations are GET, HEAD, OPTIONS, POST, PUT, PATCH, DELETE. Based on this method—and, in the case of PATCH, the parsed body—the read, append, write, and/or delete flags are assigned.

It's just parsing, really. (This shows that body should be parsed before operation though, so body can be passed as a param.)

> This makes sense, though I'm somewhat unclear on if we need any complicated parsing for the target url.

Complicated enough that node-solid-server got it wrong in a couple of places :-)

The idea is to have this in a small separate component, so we can test all parsings independently. This includes URLs with spaces, encoded characters, etc. etc.

> I feel that this belongs in each operation subclass.

We need to parse the body to determine the operation, and this is needed as input for the permissions checker. (And the operation subclasses do not [need to] know about permissions.)